### PR TITLE
Add terminology transform

### PR DESCRIPTION
**Pull request by panosk**
This is a transform for augmenting source examples with terms so the model can adapt at inference with user-provided terms.

Similar to the `inlinetags` transform, it requires a user-prepared dictionary with source-target entries, it uses the `pyahocorasick` lib for fast multi-string searches, and it brings a new, and hefty, dependency, `spacy`. Spacy is used for lemmatization as this transform works with lemmas in order to a) solve the word inflection problem when searching for terms in any form, and b) make the model inflect correctly most target terms at inference.

### CODE REVIEW

File: onmt/transforms/terminology.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+        lemmatized_source_string = " ".join(lemmatized_source)
+
+        offset = 0
+        source_with_terms = list()
+        term_counter = 0
+
+        max_terms_allowed = int(len(tokenized_source) * self.term_example_ratio)
+        is_match = False
+        for match_end, pair in self.automaton.iter_long(lemmatized_source_string):
```
Comment (francoishernandez) [inline onmt/transforms/terminology.py:107]: "pair" is not super understandable. We could naively think about it meaning "language pair".
Maybe something like (src_term, tgt_term) instead?

File: onmt/transforms/terminology.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            source_with_terms.append(lemmatized_source_string[offset:])
+            tokenized_source_with_terms = "".join(source_with_terms).split()
+
+            if not (
+                len(tokenized_source)
+                == len(lemmatized_source)
+                == len(tokenized_source_with_terms)
+            ):
+                final_string = " ".join(tokenized_source)
+                fixed_punct = re.sub(r" ([^\w\s｟\-\–])", r"\1", final_string)
```
Comment (francoishernandez) [inline onmt/transforms/terminology.py:173]: These "fixed_punct" subs look weird, why is this necessary? Side effect of the term insertion?

File: onmt/transforms/terminology.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+        )
+
+    def batch_apply(self, batch, is_train=False, stats=None, **kwargs):
+        bucket_size = len(batch)
+        examples_with_terms = 0
+
+        for i, (ex, _, _) in enumerate(batch):
+            # Skip half examples to improve performance. This means we set
+            # a hard limit for the `term_corpus_ratio` to 0.5, which is actually
+            # quite high. TODO: We can add this (skipping examples) as an option
```
Comment (francoishernandez) [inline onmt/transforms/terminology.py:325]: What kind of perf issues are we talking about if we apply this fully? (The current implementation is not super "random" so depending on how the data is looped on it might not be fully optimized augmentation.)

Comment (francoishernandez) [reply onmt/transforms/terminology.py:325]: Also, if absolutely necessary, maybe make this configurable.

**Review by francoishernandez — COMMENTED** [review]
Thanks for the contribution! A few comments below.

File: onmt/transforms/terminology.py (hunk 1 above)
Comment (panosk) [reply onmt/transforms/terminology.py:107]: If it's ok, I'll rename them to `(src_entry, tgt_entry)` as `src_term` and `tgt_term` are used later in the same method.

File: onmt/transforms/terminology.py (hunk 2 above)
Comment (panosk) [reply onmt/transforms/terminology.py:173]: Yes, indeed, these are side effects due to spacy's tokenization/lemmatization.

File: onmt/transforms/terminology.py (hunk 3 above)
Comment (panosk) [reply onmt/transforms/terminology.py:325]: > What kind of perf issues are we talking about if we apply this fully? (The current implementation is not super "random" so depending on how the data is looped on it might not be fully optimized augmentation.)

The transform adds a large overhead when every single example is passed to the term matching function and the whole training requires many workers (and lots of RAM) to keep feeding the GPU with samples. It's still under investigation but for the time being it seems to alleviate the problem.
Could you clarify a bit the part about the "randomness"?

Comment (vince62s) [reply onmt/transforms/terminology.py:325]: If we don't skip one out of two example, it makes things very slow when rebuilding a bucket so this trick is just for performance but indeed imply a hard limit of 50% max of examples to be tagged. It is not necessary to make this variable since the actual ratio is already an option. I think it is good as is.

**Comment by vince62s** [comment]
Thanks! merging
